### PR TITLE
66 save stdout contents to a file in trace mode

### PR DESCRIPTION
**Pull request by folmos-at-orange**

### CODE REVIEW

File: khiops/core/api.py [hunk 1]
```diff
@@ -112,8 +112,10 @@ def _run_task(task_name, task_args):
     task_args : dict
         Arguments of the task.
     """
-    # Save the trace argument
+    # Save the runner.run arguments other than the task parameters and options
```
Comment (popescu-v) [inline khiops/core/api.py:115]: Is this this the relevant issue associated to this PR https://github.com/KhiopsML/khiops-python/issues/66? If so, why was it unlinked from the PR? Could it be re-linked to the PR?

Comment (popescu-v) [inline khiops/core/api.py:115]: I would enclose `runner.run` in back-ticks to make it clear that this is an API fragment.

File: khiops/core/internals/runner.py [hunk 2]
```diff
@@ -612,6 +624,15 @@ def run(
             raise KhiopsRuntimeError("Khiops execution failed.") from error
         # Report any errors raised by Khiops
         else:
+            # Write the stdout and stderr streams if specified
+            if stdout_file_path is not None:
+                with open(stdout_file_path, "w") as stdout_file:
```
Comment (popescu-v) [inline khiops/core/internals/runner.py:629]: Shouldn't we allow to write these files to any location, including a remote one (like an S3 bucket etc)? In that case, shouldn't we be able to handle URIs as well? If so, perhaps the `filesystems` interface might be more appropriate?

File: khiops/core/internals/runner.py [hunk 3]
```diff
@@ -519,14 +521,24 @@ def run(
...
         general_options : `.GeneralOptions`, optional
             *Advanced:* General options for all tasks. If not set then the runner's
             values are used. Unless you know what are you doing, prefer setting this
             options with the runners accessors. See the `.GeneralOptions` class for more
             information.
+        trace : bool, default ``False``
+            If True prints the command line executed of the process and does not delete
+            any temporary files created.
+        stdout_file_path: str, optional
+            *Advanced* The path of a file to which write the stdout stream of the Khiops
```
Comment (popescu-v) [inline khiops/core/internals/runner.py:533]: I would slightly rephrase ''The path of a file to which write the stdout stream of the Khiops process." into
"If specified, write the stdout stream of the Khiops process to a file at this path."

File: khiops/core/internals/runner.py [hunk 4]
```diff
@@ -519,14 +521,24 @@ def run(
...
+            process. Normally Khiops should not write to this stream but MPI,
+            filesystems plugins or debug versions may do it. The stream is captured with
+            a UTF-8 encoding and replacing encoding errors.
+        stderr_file_path: str, optional
+            *Advanced* The path of a file to which write the stderr stream of the Khiops
```
Comment (popescu-v) [inline khiops/core/internals/runner.py:538]: I would also rephrase the first sentence as shown above.

File: khiops/core/internals/runner.py [hunk 5]
```diff
@@ -1354,6 +1375,8 @@ def raw_run(self, tool_name, command_line_args=None, use_mpi=True, trace=False):
             stdin=subprocess.DEVNULL,
             stdout=subprocess.PIPE,
             stderr=subprocess.PIPE,
+            encoding="utf8",
+            errors="replace",
             universal_newlines=True,
         ) as khiops_process:
             stdout, stderr = khiops_process.communicate()
```
Comment (popescu-v) [inline khiops/core/internals/runner.py:1382]: Shouldn't we add or update some core tests according to this change (perhaps add a test to `KhiopsCoreIOTests`)?

**Review by popescu-v — CHANGES_REQUESTED** [review]

File: khiops/core/api.py (hunk 1 above)
Comment (folmos-at-orange) [reply khiops/core/api.py:115]: No but there won't be an issue to improve a comment.

File: khiops/core/internals/runner.py [hunk 6]
```diff
@@ -612,6 +624,15 @@ def run(
             raise KhiopsRuntimeError("Khiops execution failed.") from error
         # Report any errors raised by Khiops
         else:
+            # Write the stdout and stderr streams if specified
+            if stdout_file_path is not None:
```
[the remark this answers is not in the document]
Comment (folmos-at-orange) [reply khiops/core/internals/runner.py:628]: No. We do not do this kind of check anywhere in the library. If there aren't permissions then the good "permission denied" exception will be raised by Python.

File: khiops/core/internals/runner.py (hunk 2 above)
Comment (folmos-at-orange) [reply khiops/core/internals/runner.py:629]: Yes. I will refactor to use `fs`

File: khiops/core/internals/runner.py (hunk 3 above)
Comment (folmos-at-orange) [reply khiops/core/internals/runner.py:533]: It is already implied that it is optional. I will change the signature and comment. to match that of `log_file_path`

File: khiops/core/internals/runner.py (hunk 4 above)
Comment (folmos-at-orange) [reply khiops/core/internals/runner.py:538]: See above.

File: khiops/core/internals/runner.py (hunk 5 above)
Comment (folmos-at-orange) [reply khiops/core/internals/runner.py:1382]: I thought about it but it requires heavy mocking, because we need to simulate the `stdout` and `stderr` of the Khiops process.

File: khiops/core/internals/runner.py (hunk 3 above)
Comment (popescu-v) [reply khiops/core/internals/runner.py:533]: Ok.

File: khiops/core/internals/runner.py (hunk 4 above)
Comment (popescu-v) [reply khiops/core/internals/runner.py:538]: Ok.

File: khiops/core/internals/runner.py (hunk 5 above)
Comment (popescu-v) [reply khiops/core/internals/runner.py:1382]: Or we could have it as an integration test (w/o mocking) for the time being, with a TODO in the commit message that this should be mocked to make it into a unit test?

Comment (folmos-at-orange) [reply khiops/core/internals/runner.py:1382]: And I'll make an issue for the mocking test.

File: khiops/core/api.py (hunk 1 above)
Comment (folmos-at-orange) [reply khiops/core/api.py:115]: Done

File: khiops/core/internals/runner.py [hunk 7]
```diff
@@ -524,9 +526,19 @@ def run(
...
             *Advanced:* General options for all tasks. If not set then the runner's
-            values are used. Unless you know what are you doing, prefer setting this
-            options with the runners accessors. See the `.GeneralOptions` class for more
-            information.
+            values are used. Unless you know what are you doing, prefer setting these
+            options with the runner's accessors. See the `.GeneralOptions` class for
+            more information.
+        stdout_file_path : str, optional
+            *Advanced* If specified it writes the stdout stream of the Khiops process
+            that path. Normally Khiops should not write to this stream but MPI,
```
Comment (popescu-v) [inline khiops/core/internals/runner.py:534]: s/"that path"/"to that path"/

File: khiops/core/internals/runner.py [hunk 8]
```diff
@@ -524,9 +526,19 @@ def run(
...
+            filesystems plugins or debug versions may do it. The stream is captured with
+            a UTF-8 encoding and replacing encoding errors.
+        stderr_file_path : str, optional
+            *Advanced* If specified it writes the stderr stream of the Khiops process
```
Comment (popescu-v) [inline khiops/core/internals/runner.py:538]: s/"that path"/"to that path"/

**Review by popescu-v — CHANGES_REQUESTED** [review]
Some minor comments.

File: khiops/core/internals/runner.py (hunk 7 above)
Comment (folmos-at-orange) [reply khiops/core/internals/runner.py:534]: Ok

File: khiops/core/internals/runner.py (hunk 8 above)
Comment (folmos-at-orange) [reply khiops/core/internals/runner.py:538]: Ok

File: doc/notes.rst [hunk 9]
```diff
@@ -1,25 +1,101 @@
...
+    Path of the output Khiops scenario file (command line option ``-o`` of the desktop app).  If
+    the empty string is specified no output scenario file is generated.
+task_file_path : str, default ""
+    Path of the task file (command line option ``-p`` of the desktop app). If the empty
+    string is specified no task file is generated.
+trace : bool, default ``False``
+    If True prints the command line executed of the process and does not delete any temporary files
+    created.
+stdout_file_path : str, default ""
+    *Advanced* Path of a file to which write the Khiops process stdout stream.  Normally Khiops
```
Comment (popescu-v) [inline doc/notes.rst:24]: I would rephrase the first sentence to "Path of a file the Khiops process writes the stdout stream to".

File: doc/notes.rst [hunk 10]
```diff
@@ -1,25 +1,101 @@
...
+    should not write to this stream but MPI, filesystems plugins or debug versions may do it. The
+    stream is captured with a UTF-8 encoding and replacing encoding errors. If equal to "" then it
+    writes no file.
+stderr_file_path : str, default ""
+    *Advanced* Path of a file to which write the Khiops process stderr stream.  Normally Khiops
```
Comment (popescu-v) [inline doc/notes.rst:29]: I would rephrase the first sentence (cf. the `stdout`-specific sentence above).

File: khiops/core/dictionary.py [hunk 11]
```diff
@@ -88,9 +88,9 @@ class DictionaryDomain(KhiopsJSONObject):
         specified it returns an empty instance.
 
         .. note::
-            Prefer either the `read_khiops_dictionary_json_file` method or the
-            `.read_dictionary_file` function from the core API to obtain an instance of
-            this class from a Khiops Dictionary file (``kdic`` or ``kdicj``).
+            Prefer either the the `.read_dictionary_file` function from the core API to
```
Comment (popescu-v) [inline khiops/core/dictionary.py:91]: Delete "either the".

File: khiops/core/internals/runner.py [hunk 12]
```diff
@@ -524,9 +528,19 @@ def run(
...
         general_options : `.GeneralOptions`, optional
             *Advanced:* General options for all tasks. If not set then the runner's
-            values are used. Unless you know what are you doing, prefer setting this
-            options with the runners accessors. See the `.GeneralOptions` class for more
-            information.
+            values are used. Unless you know what are you doing, prefer setting these
+            options with the runner's accessors. See the `.GeneralOptions` class for
+            more information.
+        stdout_file_path : str, default ""
+            *Advanced* Path of a file to which write the Khiops process stdout stream.
```
Comment (popescu-v) [inline khiops/core/internals/runner.py:535]: I would rephrase the first sentence (see relevant comment on the associated .rst documentation file).

File: khiops/core/internals/runner.py [hunk 13]
```diff
@@ -524,9 +528,19 @@ def run(
...
+            Normally Khiops should not write to this stream but MPI, filesystems plugins
+            or debug versions may do it. The stream is captured with a UTF-8 encoding
+            and replacing encoding errors. If equal to "" then it writes no file.
+        stderr_file_path : str, default ""
+            *Advanced* Path of a file to which write the Khiops process stderr stream.
```
Comment (popescu-v) [inline khiops/core/internals/runner.py:540]: I would rephrase the first sentence (see relevant comments above). Namely, "to which write" needs to be rephrased IMHO.

File: tests/test_core.py [hunk 14]
```diff
@@ -653,14 +657,129 @@ def test_runner_version(self):
...
+    def test_std_stream_warnings(self):
+        def _test(stream_name, mock_raw_run):
+            # Create the function arguments
+            stream_args = ["a.kdic", "a", "a.txt"]
+
+            # Run a normal local runner with the `raw_run` function mocked
+            default_runner = kh.get_runner()
+            mock_runner = KhiopsLocalRunner()
+            mock_runner._initialize_khiops_environment()
+            kh.set_runner(mock_runner)
```
Comment (popescu-v) [inline tests/test_core.py:717]: Perhaps this snippet can be factored into a helper function, as it seems repetitive?

**Review by popescu-v — CHANGES_REQUESTED** [review]

File: doc/notes.rst (hunk 9 above)
Comment (folmos-at-orange) [reply doc/notes.rst:24]: What about `Path to a file where the Khiops process writes its stdout stream`

File: khiops/core/dictionary.py (hunk 11 above)
Comment (folmos-at-orange) [reply khiops/core/dictionary.py:91]: Ok

File: tests/test_core.py (hunk 14 above)
Comment (folmos-at-orange) [reply tests/test_core.py:717]: I don't know. It changes a global state (`set_runner`)

File: khiops/core/internals/runner.py (hunk 13 above)
Comment (folmos-at-orange) [reply khiops/core/internals/runner.py:540]: See above.

File: khiops/core/internals/runner.py (hunk 12 above)
Comment (folmos-at-orange) [reply khiops/core/internals/runner.py:535]: See above.

File: doc/notes.rst (hunk 10 above)
Comment (folmos-at-orange) [reply doc/notes.rst:29]: See above

Comment (popescu-v) [reply doc/notes.rst:29]: Ok.

File: khiops/core/internals/runner.py (hunk 12 above)
Comment (popescu-v) [reply khiops/core/internals/runner.py:535]: Ok.

File: khiops/core/internals/runner.py (hunk 13 above)
Comment (popescu-v) [reply khiops/core/internals/runner.py:540]: Ok.

File: tests/test_core.py (hunk 14 above)
Comment (popescu-v) [reply tests/test_core.py:717]: Indeed, it would be a side-effectful function. I don't see any particular issue with this in a testing context, as long as its side effects are documented (docstring + comment).

File: doc/notes.rst (hunk 9 above)
Comment (folmos-at-orange) [reply doc/notes.rst:24]: Done

File: doc/notes.rst (hunk 10 above)
Comment (folmos-at-orange) [reply doc/notes.rst:29]: Done

File: khiops/core/dictionary.py (hunk 11 above)
Comment (folmos-at-orange) [reply khiops/core/dictionary.py:91]: Done

File: khiops/core/internals/runner.py (hunk 12 above)
Comment (folmos-at-orange) [reply khiops/core/internals/runner.py:535]: Done

File: khiops/core/internals/runner.py (hunk 13 above)
Comment (folmos-at-orange) [reply khiops/core/internals/runner.py:540]: Done

File: tests/test_core.py (hunk 14 above)
Comment (folmos-at-orange) [reply tests/test_core.py]: Done

**Review by popescu-v — APPROVED** [review]
